scripts/cosmosdb_migration.py

```diff
@@ -0,0 +1,162 @@
+        self.client = None
+        self.database = None
+        self.old_container = None
+        self.new_container = None
+
+    async def connect(self):
+        """
+        Connect to CosmosDB and initialize containers.
+        """
+        self.client = CosmosClient(
+            url=f"https://{self.cosmos_account}.documents.azure.com:443/", credential=self.credential
+        )
+        self.database = self.client.get_database_client(self.database_name)
+        self.old_container = self.database.get_container_client("chat-history")
+        self.new_container = self.database.get_container_client("chat-history-v2")
+        try:
+            await self.old_container.read()
+        except Exception:
+            raise ValueError(f"Old container {self.old_container.id} does not exist")
+        try:
+            await self.new_container.read()
+        except Exception:
+            raise ValueError(f"New container {self.new_container.id} does not exist")
+
+    async def migrate(self):
+        """
+        Migrate data from old schema to new schema.
+        """
+        if not self.client:
+            await self.connect()
+        if not self.old_container or not self.new_container:
+            raise ValueError("Containers do not exist")
+
+        query_results = self.old_container.query_items(query="SELECT * FROM c")
+
+        item_migration_count = 0
+        async for page in query_results.by_page():
+            async for old_item in page:
+                batch_operations = []
+                # Build session item
+                session_item = {
+                    "id": old_item["id"],
+                    "version": "cosmosdb-v2",
+                    "session_id": old_item["id"],
+                    "entra_oid": old_item["entra_oid"],
+                    "title": old_item.get("title"),
+                    "timestamp": old_item.get("timestamp"),
+                    "type": "session",
+                }
+                batch_operations.append(("upsert", (session_item,)))
+
+                # Build message_pair
+                answers = old_item.get("answers", [])
+                for idx, answer in enumerate(answers):
+                    question = answer[0]
+                    response = answer[1]
+                    message_pair = {
+                        "id": f"{old_item['id']}-{idx}",
+                        "version": "cosmosdb-v2",
+                        "session_id": old_item["id"],
+                        "entra_oid": old_item["entra_oid"],
+                        "type": "message_pair",
+                        "question": question,
+                        "response": response,
+                        "order": idx,
+                        "timestamp": None,
+                    }
+                    batch_operations.append(("upsert", (message_pair,)))
+
+                # Execute the batch using partition key [entra_oid, session_id]
```

Comment:
comment is wrong? we are using [entra_oid, id]

Reply:
The old item represented the session, so old_item["id"] gives us the session_id